tool/templates/js/runtime.mjs

```diff
@@ -690,3 +690,66 @@ export class GarbageCollectorGrip {
 }
 
 const DiplomatBufferFinalizer = new FinalizationRegistry(free => free());
+
+/**
+ * For allocating and cleaning up structs to be passed into the C Spec WASM ABI.
+ *
+ * Created when we load in the WebAssembly.
+ */
+export class FunctionParamAllocator {
+    #wasm = null;
+
+    #ptr = 0;
+    #capacity = 0;
+
+    #allocated = [];
+    #currentPtr = 0;
+
+    init(symbol, wasm) {
+        if (symbol === internalConstructor) {
+            this.#wasm = wasm;
+
+            // Hack, not sure else how to handle things though:
+            if (this.#ptr === 0){
+                this.#ptr = this.#wasm.diplomat_alloc(this.#capacity, 1);
+            }
+        }
+    }
+
+    // Called in the last line of this file, call should be generated by diplomat.
+    reserve(capacity) {
+        if (this.#ptr !== 0) {
+            throw new Error("FunctionParamAllocator already reserved.");
+        }
+
+        this.#capacity = capacity;
+    }
+
+    alloc(size) {
```

Comment:
nit: please document this and the other function

question: this returns a relative pointer (an index) whereas `.pop` returns an absolute pointer.

We should be consistent. I think we should consistently store indices, and return absolute pointers.